### ISSUE TITLE
Optional thumbnail image inside .glb file

### ISSUE THREAD

**Issue by Arlorean** [KhronosGroup/glTF#1167]
Provide an option to have an embedded thumbnail file (.png or .jpg) in the .glb file.
Windows Explorer can then show that thumbnail, making it easier to find assets on disk.

Ideally the **.glb** file would just be a **.zip** format file so you could store **thumbnail.png** inside the file, together with other information like the author, etc. They would just go along with the .glb (.zip format) file.

If we're too late to make .glb a .zip file, how about we create a new file extension, **.glz** that is a .zip format file containing the **.gltf,** any **textures** in their original file format inside the .zip, etc. It would be exactly the same as the current .zip file that we can download from _SketchFab_ but with the .glz extension and optionally containing a thumbnail.png file.

**Comment by donmccurdy** [comment]
There's some discussion of thumbnails in [this thread](https://github.com/KhronosGroup/glTF/issues/1051), although I think having this separate issue for the question is better.

My own preference would be what @CentaurMare describes, where the glTF file itself does not contain the thumbnail.

> In Windows, the explorer shell looks up the registry for the file type and if present calls the mapped DLL via the IThumbnailProvider interface - it would be this DLL (which someone needs to write) that creates the thumbnail so that the Windows Explorer can update its thumbnail cache.
This means tweaking a value (such as a colour) directly in the glTF file will cause its thumbnail to be regenerated, which is what we want.

Similarly, MacOS can generate thumbnails and [full 3D previews](https://github.com/magicien/GLTFQuickLook) without changes to the file spec; it does this for OBJ already.

In any case, `.glb` is not a zip format, but can easily contain image files regardless. Adding a thumbnail via extension or `.extras` property would be easy to do within the existing specification. My vote would be to leave it ad-hoc, though, rather than try to make thumbnails part of the specification and increase the size of files that are intended to be runtime-focused.

**Comment by Arlorean** [comment]
Thanks for the link to the other discussion and for the quick reply.

I agree, I'd rather not see the thumbnail in the **.gltf** file itself, but I would be keen to see it as an optional external reference, like textures.

The **.glb** looks so close to the **.zip** format that I wondered if we could just use .zip format to make it easier to use existing tool-chains to create, view and extract/replace content files. This is similar to the way Microsoft have setup Office file formats **.pptx**, **.docx**, etc. as **.zip** files (with thumbnails), and also their .**3mf** file format. Zip file entries can be stored uncompressed so they can be processed just as quickly as they can from a .glb as well. For the same reason you chose the JSON standard for the manifest format, I would have chosen ZIP for the package format, standards and tooling reuse.

Explorer shell extensions can be used to create thumbnails on the fly (I've done it myself for [Voxel Files](https://github.com/Arlorean/Voxels)) but it can take some time to create these thumbnails and lower end machines may struggle with larger assets. Extracting a small thumbnail from a .zip is very fast and doesn't require any extra libraries that aren't already on most machines.

> My vote would be to leave it ad-hoc, though, rather than try to make thumbnails part of the specification and increase the size of files that are intended to be runtime-focused

This is the best argument for not doing it I agree.

However, the downloads from Sketchfab are often >10Mb per asset and a small thumbnail would only add a few Kb to that file. Plus, it can still be optional if the file is only ever used at runtime. Deleting the thumbnail can be done using 7-zip or a similar tool.

**Comment by space2** [comment]
IMHO even a 1-2KB thumbnail is very useful. For example I couldn't get the OSX Quick View plugin working (maybe because I'm still using 10.9?). And it's much simpler to write a plugin to extract a simple thumbnail then to extract, parse and render the mesh.

So I'm in favor of having a thumbnail in the glb file. The simplest solution would actually be by adding a new (3rd) chunk to the glb. In that case the thumbnail could be extract with ~20 lines of code, without using any json parsing.

**Comment by Arlorean** [comment]
> The simplest solution would actually be by adding a new (3rd) chunk to the glb. In that case the thumbnail could be extract with ~20 lines of code, without using any json parsing.

That seems like the most practical solution to me too, although I'm still keen on the .zip idea as you can probably tell. :)

**Comment by emackey** [comment]
The .zip idea is nice for re-use of existing tooling, but, some engines ingesting GLB don't want the legacy of zip.  There are a number of efforts underway now to add compression, but these are GPU compression (Draco mesh, Basis textures) not CPU compression (zip).  The difference is, the asset could remain compressed on the GPU (or be transpiled to local-compatible GPU compression) while it is in runtime use, which zip can't do.  Zip only helps the network, not the GPU.

**Comment by lexaknyazev** [comment]
Just for the record: mesh data stays uncompressed on the GPU (i.e. Draco is CPU-compression). Nevertheless, domain-specific compression routines are usually more efficient than general-purpose approaches, so we expect glTF to have standardized texture and geometry codecs at some point. Applying ZIP on top of that likely won't provide much value.

**Comment by zellski** [comment]
(And, at least in the domain of WebVR, browser transmission is of course nearly invariably gzipped anyway, so another layer of unsolicited compression is actually unwelcome.)

**Comment by donmccurdy** [comment]
For OS support (e.g. seeing a thumbnail on your desktop), I think thumbnails must be handled by the OS long-term, and we shouldn't try to shortcut that by adding things to the file. We really want the OS to show a 3D preview as MacOS does for OBJ and COLLADA, not a thumbnail, in the long run.

For other use cases like APIs, I would refer to the new Poly API. It defines metadata like thumbnails through a higher-level abstraction as described here: https://developers.google.com/poly/reference/api/rest/v1/assets#Asset. I would encourage other APIs and web-based projects to do the same.

The question of ZIP compression is separate, and wouldn't be needed for thumbnails anyway. As others have explained, I think current `.glb` and upcoming compression techniques like Draco will be most efficient for runtime needs.

**Comment by emackey** [comment]
> Also note there are tools in progress now for easily zipping and un-zipping .glb assets

The one that was originally contributed to the VSCode extension has since been broken out into an npm package:

https://www.npmjs.com/package/gltf-import-export

**Comment by Arlorean** [comment]
It's interesting to note that the new Apple/Pixar format [USDZ](https://graphics.pixar.com/usd/docs/Usdz-File-Format-Specification.html) is a **.zip** package **without compression**. I still think that's a great way to store gltf content. I haven't seen a reference to a USDZ thumbnail yet but if anyone does I'd love to know about it and I'll bet it's a PNG/JPG in the ZIP.

I'm happy for this issues to be _closed_ if there's no sign of concensus on adding thumbnail support BTW.

**Comment by CentaurMare** [comment]
Looking at a .3mf (.zip) file, the thumbnail is embedded within the file. This allows the thumbnail provider to simply read the file and quickly extract the image. This can't be done with a glTF file, as it could take some time for a large model to load and render, and we can't guarantee that an exporter will create a thumbnail.
I'm considering building a Windows thumbnail provider for .gltf and .glb files which will read the thumbnail, not from the file itself, but from an NTFS Additional Data Stream, e.g. myModel.glb:thumbnail
Then have an indexing type of service that scans for and generates thumbnails from the glTF files on the system and writes them into the thumbnail stream.
Advantages:
The thumbnail isn't part of the file, it is not sent across the network, saving bandwidth.
The thumbnail is moved with the file if the file is relocated on disk.
Disadvantages:
The thumbnail is lost if copied to a FAT32 device such as a USB.
A thumbnail would not be generated if the file was read-only.

The issue I'm facing at the moment is that my glTF viewer requires a window to render into (It is written in C# with OpenTK). I can then save the render as an image, but it still brings up the Window. Does anyone have a glTF viewer capable of rendering to an image file, or know how to suppress the window creation with OpenTK?

Or can someone chime in with a better plan for thumbnails?

**Comment by gkv311** [comment]
CentaurMare wrote:
> or know how to suppress the window creation with OpenTK?

There is no need displaying the window - if you keep it hidden it should work OK - see comments on this thread (though make sure you are using FBO for image dump to avoid undefined behavior):
https://github.com/opentk/opentk/issues/459

From another side, I would say that using OpenGL might be not very good idea on Windows for making thumbnails. I'm not familiar with Windows thumbnail provider logic, but would expect it to be performed preferably in background, when user is inactive, and in this context OpenGL-based application might observe issues with no hardware-accelerated OpenGL driver being available.

Another issue is that glTF file might be really huge, and there are no tricks similar to many image formats to be loaded only partially (fitting in desired memory limits), so that this process should consider somehow how to make screenshots without slowing down foreground system processes / making system unstable.

**Comment by CentaurMare** [comment]
I tried the hidden window approach, but the window was first created visible, then hidden. I could render in GDI+ using only material colours - so ignore textures. That would be much slower to create the render than using OpenGL but it would work as a background service, and if it was a background service with a FileSystemWatcher it could chug away in the background unobtrusively and create thumbnails as needed. I might give that a go and see how the thumbnails look, but I know already they wont look as good as the OpenGL approach. But at least it represents the model.

I've got the IThumbnailProvider handler working, it simply loads the thumbnail from the ":thumbnail" stream attached to the file. It uses the IInitializeWithFile interface to get the filename, appends ":thumbnail" and reads in the image data (PNG or JPG).

**Comment by CentaurMare** [comment]
@gkv311

> I would say that using OpenGL might be not very good idea on Windows for making thumbnails

So I went with your suggestion, and it works nicely!

My thumbnail generator is a C# application.

I'm rendering by z-sorting each triangle and drawing using Graphics.FillPolygon, using the material colour and calculating incoming light based on the normal. I imagined it would be horribly slow to render, but it turns out to be fairly quick.

I'm rendering to a 4096 x 4096 bitmap and then resampling to 1024 x 1024, then saving the bitmap
as a PNG into the ":thumbnail" alternate data stream on the file.

The C++ Thumbnail Provider simply reads from that stream to present the thumbnail to Windows Explorer (which downsizes appropriately)

Some things I need to resolve:

  Thumbnail generation
    My Thumbnail generator will need to run as a service.
    To resolve synchronization issues, the ideal would be to have the thumbnail provider signal the thumbnail generator service that it needs a thumbnail on a file, which is then generated and returned to the thumbnail provider. That way there is no need for an intermediate ":thumbnail" stream, and Windows can handle storage of the thumbnail in Thumbs.db
	
  Graphics
    I will need to calculate the normals if they are not provided.
    For textured models, I have no clue how to render the texture. I could sample a pixel from the texture and use that as the material colour if all else fails.
	It also ideally needs Gouraud shading, but I don't know how to do this.
	I need to use a Perspective Transform.
	I won't bother with all the other graphics features that an OpenGL fragment shader can provide.

**Comment by CentaurMare** [comment]
Success! (mostly)

I attempted to use PathGradientBrush to do Gouraud shading, but it seems there is a memory leak in Microsoft's implementation and it quickly ran out of memory. So the resulting image is somewhat faceted, but otherwise it represents the model.

I'm sampling the texture at each point in the model, and using an average colour for the face. It works if the model is high polygon.

I've added the emissive material factor, but metallic-roughness will need to be left out.

I'm no longer using the :thumbnail stream, instead the thumbnail provider (C++) requests the image from the thumbnail service (C#) via named pipes. So updating the glTF file will mean the thumbnail gets regenerated as expected.


Prerequisites: .NET Framework 4.7.2 (Probably 4.x will work)
Requires 64-bit Windows, I'm assuming no one is using an ancient PC for graphics work.

I'm not supporting Draco Mesh Compression, but otherwise let me know if a model thumbnail won't view and I will take a look at why.

[ThumbnailProvider.zip](https://github.com/KhronosGroup/glTF/files/2683916/ThumbnailProvider.zip)
[ThumbnailProviderService.zip](https://github.com/KhronosGroup/glTF/files/2683917/ThumbnailProviderService.zip)

**Comment by ErwinMeulman** [comment]
What's going wrong???????????????????????


Windows Registry Editor Version 5.00

[HKEY_CLASSES_ROOT\.glb\shellex\{125056E5-A8A3-41F8-A5F8-B834ABDEEFD9}]
@="{6F6A251D-E800-4870-B088-87FA9B42F685}"

[HKEY_CLASSES_ROOT\.gltf\shellex\{125056E5-A8A3-41F8-A5F8-B834ABDEEFD9}]
@="{6F6A251D-E800-4870-B088-87FA9B42F685}"

﻿[HKEY_CLASSES_ROOT\glTFThumbnailProvider]
@="glTFThumbnailProvider"

[HKEY_CLASSES_ROOT\glTFThumbnailProvider\CLSID]
@="{6F6A251D-E800-4870-B088-87FA9B42F685}"

[HKEY_CLASSES_ROOT\WOW6432Node\CLSID\{6F6A251D-E800-4870-B088-87FA9B42F685}]
@="glTFThumbnailProvider"
"DisplayName"="glTF Thumbnail Provider"
"AppID"="{AF6BF7B1-6567-4D3F-A0FA-D938C7D85A8A}"
"AllowEdit"=dword:00000000
"DisableLowILProcessIsolation"=dword:00000001

[HKEY_CLASSES_ROOT\WOW6432Node\CLSID\{6F6A251D-E800-4870-B088-87FA9B42F685}\Implemented Categories\{E073B408-1A6A-4B0A-BF49-2A47FEE7939E}]

[HKEY_CLASSES_ROOT\WOW6432Node\CLSID\{6F6A251D-E800-4870-B088-87FA9B42F685}\InprocServer32]
@="mscoree.dll"
"ThreadingModel"="Both"
"Class"="glTFThumbnailProvider"
"Assembly"="glTFThumbnailProvider, Version=1.5.0.0, Culture=neutral, PublicKeyToken=null"
"RuntimeVersion"="v4.0.30319"
"CodeBase"="file:///c:/program files/glTF Thumbnail Provider/glTFThumbnailProvider.dll"

[HKEY_CLASSES_ROOT\WOW6432Node\CLSID\{6F6A251D-E800-4870-B088-87FA9B42F685}\InprocServer32\1.5.0.0]
"Class"="glTFThumbnailProvider"
"Assembly"="glTFThumbnailProvider, Version=1.5.0.0, Culture=neutral, PublicKeyToken=null"
"RuntimeVersion"="v4.0.30319"
"CodeBase"="file:///c:/program files/glTF Thumbnail Provider/glTFThumbnailProvider.dll"

[HKEY_CLASSES_ROOT\WOW6432Node\CLSID\{6F6A251D-E800-4870-B088-87FA9B42F685}\ProgId]
@="glTFThumbnailProvider"






[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\PreviewHandlers]
"{6F6A251D-E800-4870-B088-87FA9B42F685}"="glTF Thumbnail Provider"

[HKEY_LOCAL_MACHINE\SOFTWARE\Classes\glTFThumbnailProvider]
@="glTFThumbnailProvider"

[HKEY_LOCAL_MACHINE\SOFTWARE\Classes\glTFThumbnailProvider\CLSID]
@="{6F6A251D-E800-4870-B088-87FA9B42F685}"

[HKEY_LOCAL_MACHINE\SOFTWARE\Classes\Wow6432Node\CLSID\{6F6A251D-E800-4870-B088-87FA9B42F685}]
@="glTFThumbnailProvider"
"AllowEdit"=dword:00000000
"AppID"="{AF6BF7B1-6567-4D3F-A0FA-D938C7D85A8A}"
"DisableLowILProcessIsolation"=dword:00000001
"DisplayName"="glTF Thumbnail Provider"

[HKEY_LOCAL_MACHINE\SOFTWARE\Classes\Wow6432Node\CLSID\{6F6A251D-E800-4870-B088-87FA9B42F685}\Implemented Categories]

[HKEY_LOCAL_MACHINE\SOFTWARE\Classes\Wow6432Node\CLSID\{6F6A251D-E800-4870-B088-87FA9B42F685}\Implemented Categories\{E073B408-1A6A-4B0A-BF49-2A47FEE7939E}]

[HKEY_LOCAL_MACHINE\SOFTWARE\Classes\Wow6432Node\CLSID\{6F6A251D-E800-4870-B088-87FA9B42F685}\InprocServer32]
@="mscoree.dll"
"Assembly"="glTFThumbnailProvider, Version=1.5.0.0, Culture=neutral, PublicKeyToken=null"
"Class"="glTFThumbnailProvider"
"CodeBase"="file:///c:/program files/glTF Thumbnail Provider/glTFThumbnailProvider.dll"
"RuntimeVersion"="v4.0.30319"
"ThreadingModel"="Both"

[HKEY_LOCAL_MACHINE\SOFTWARE\Classes\Wow6432Node\CLSID\{6F6A251D-E800-4870-B088-87FA9B42F685}\InprocServer32\1.5.0.0]
"Assembly"="glTFThumbnailProvider, Version=1.5.0.0, Culture=neutral, PublicKeyToken=null"
"Class"="glTFThumbnailProvider"
"CodeBase"="file:///c:/program files/glTF Thumbnail Provider/glTFThumbnailProvider.dll"
"RuntimeVersion"="v4.0.30319"

[HKEY_LOCAL_MACHINE\SOFTWARE\Classes\Wow6432Node\CLSID\{6F6A251D-E800-4870-B088-87FA9B42F685}\ProgId]
@="glTFThumbnailProvider"






[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Installer\UserData\S-1-5-18\Components\4BEB160A044200649A098D1938ABC7CD]
"62A625FC464230000000209E05916B82"="Cc:\\program files\\glTF Thumbnail Provider\\glTFThumbnailProvider.dll"

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Installer\UserData\S-1-5-18\Components\4BEB160A144200649A098D1938ABC7CD]
"62A625FC464230000000209E05916B82"="c:\\program files\\glTF Thumbnail Provider\\glTFThumbnailProvider.dll"

**Comment by CentaurMare** [comment]
Hi Erwin,

Some things look very weird...
"Cc:\program files\glTF Thumbnail Provider\glTFThumbnailProvider.dll"
Should start with "C:\ not "Cc:\

and "file:///c:/program files/glTF Thumbnail Provider/glTFThumbnailProvider.dll" is not a windows path.

You are installing on Windows, right?

So, step one, make sure you are logged in as an administrator, as we are installing a windows service.
Step two, ensure .net framework 4.x is installed

On a command prompt, type the following...
cd %windir%\Microsoft.NET\FrameWork
dir

you should see a directory called v4.0.30319

The service is actually targetted for .net framework 4.7.2, so if you don't have that installed, download it from microsoft.

Check the contents of "c:\program files\glTF Thumbnail Provider", it has to be unzipped there, not anywhere else. It should contain at least the following files.

Centauri.Model.Gltf.dll
Centauri.Model.Gltf.Interface.dll
Centauri.Model.Gltf.Load.dll
Centauri.Model.Gltf.Load.Json.dll
Centauri.Model.Gltf.Thumbnail.dll
Centauri.Model.Gltf.Utilities.dll
glTFThumbnailProvider.dll
glTFThumbnailProviderService.exe
install.cmd
OpenTK.dll
uninstall.cmd



Then, on the command prompt, navigate to 
cd "c:\program files\gltf thumbnail provider"
install

What is the output from running install?

**Comment by ErwinMeulman** [comment]
I wan't to register the glTFThumbnailProvider.dll, so I can preview it in Outlook. But the register I copied is for a c# dll and the glTFThumbnailProvider is a c++, als I don't now the AppId. Is the register for glTFThumbnailProvider the same or do I need a other way?

[Preview_Register.zip](https://github.com/KhronosGroup/glTF/files/2847768/Preview_Register.zip)

![compare](https://user-images.githubusercontent.com/22874090/52520441-8a4d7b80-2c6a-11e9-924c-3cb54ced2e4a.png)

![09-02-2019 11-50-26](https://user-images.githubusercontent.com/22874090/52520435-7dc92300-2c6a-11e9-8f7f-e7002058e293.png)

![previewhandler](https://user-images.githubusercontent.com/22874090/52520432-743fbb00-2c6a-11e9-8a4c-fdb8041fd449.png)

This is that I wan't
![outlook_preview_dwf_pdf_glb](https://user-images.githubusercontent.com/22874090/52520427-6853f900-2c6a-11e9-8c12-b484f502c632.gif)

**Comment by CentaurMare** [comment]
@ErwinMeulman

Hi Erwin,

The thumbnail provider only provides thumbnails (icons) for Windows Explorer, it can't display an interactive model that you can rotate etc from within outlook. I do have a separate application to load and display a .glTF model, but it is standalone and not embedded within another application such as outlook.

I might investigate what is required for outlook, but no promises :)

**Comment by Arlorean** [comment]
Great job with the thumbnail viewer @CentaurMare. Very impressive. It's a shame there isn't a reference renderer that could be used as a separate process that we could call out to to render these thumbnails. Maybe there is?

I just noticed that [Kenney.nl](https://kenney.nl/) has released GLTF versions of his free assets. They provide some very simple models that highlight issues with missing (or likely reversed) faces that aren't being rendered currently in the thumbnails.

Try out [Kenney's Weapon Pack](https://kenney.nl/assets/weapon-pack) to see the issues. Here are the assets on [SketchFab](https://sketchfab.com/3d-models/weapon-pack-game-assets-8507b9f830154e8286e0a01980050f73) and here they are in Windows File Explorer as thumbnails:
![image](https://user-images.githubusercontent.com/1547800/53188840-16ae5580-35fe-11e9-8444-c825feaa7ff0.png)

**Comment by donmccurdy** [comment]
> It's a shame there isn't a reference renderer that could be used as a separate process that we could call out to to render these thumbnails. Maybe there is?

There is one in development, although to be clear it is really meant as a "reference", and for efficiently generating hundreds of thumbnails you may have better luck with a made-for-production renderer. ThreeJS and Filament can be used via CLI (and probably BabylonJS?). There are [other options](https://github.com/KhronosGroup/glTF#loaders-and-engines), too, but I haven't used many of them myself.

**Comment by CentaurMare** [comment]
@Arlorean,

I am unable to download the weapons pack as individual glb files, did you import the wavefront .obj files into an application and then export to .glb? If you are able to upload an individual .glb, I will be able to determine if my thumbnail generator is at fault - you may have installed the earlier version?
A triangle will not be drawn if it is facing away from the viewer and its material is not double-sided. This is intended as some models use this to great effect, such as for drawing black outlines around a model.

>  It's a shame there isn't a reference renderer that could be used as a separate process that we could call out to to render these thumbnails. Maybe there is?

Funny thing is, that is exactly how my thumbnail generator works - it passes the name of the .gltf or .glb file via a named-pipe message to the thumbnail generator running as a windows service, which loads and generates the image, and then sends the resulting .png image data back to the thumbnail provider.
(The thumbnail provider is a C++ application, and the renderer is a C# application)

I also have a glTF viewer which renders beautifully using OpenGL shaders (it even does the animations), and I had originally intended to use that as the renderer because it can render with PBR materials etc - but the issue I was running into is that the OpenTK library doesn't allow me to render in the background, it was designed for renderering into a window.

Anyway, I'm really pleased to hear that someone finds the thumbnail generator useful! I didn't know if anyone was using it at all. So thanks for the feedback!

**Comment by Arlorean** [comment]
> I didn't know if anyone was using it at all. So thanks for the feedback!

It is definitely useful. Thanks for doing it. I did try it when you first announced it but didn't get around to replying to you. Sorry about that. Do you have a link to the latest version I can try out?

Github doesn't let me upload a **.glb** file individually so I've zipped them up and attached them here. Kenney's assets are all low poly, generally without textures, so they should be a simple test case:
[GLTF format.zip](https://github.com/KhronosGroup/glTF/files/2896795/GLTF.format.zip)

Here's the **rockerlauncher.glb** file in Windows 10 3D Viewer:
![image](https://user-images.githubusercontent.com/1547800/53286532-2053cc80-3768-11e9-9f31-46fc07263383.png)

Here's the generated thumbnail:
![image](https://user-images.githubusercontent.com/1547800/53286536-395c7d80-3768-11e9-9e92-b3275a846498.png)

**Comment by CentaurMare** [comment]
[cid:3e3b95bd-46d7-4201-aad7-17ea0067c063]

I've reattached the latest version, stop the glTF Thumbnail Provider service first from the control panel, unzip into c:\program files\glTF Thumbnail Provider
and then restart the service.
The visual artifacts that are visible on the ammo are caused by the z-sort of the facets before drawing, this won't be an easy fix without a z-buffer, so I might need to have a go at writing my own triangle-drawing routine with z-buffer - visually it would look better but then it would also be slower. Unless I can get OpenTK running without a window.


________________________________

**Comment by CentaurMare** [comment]
Try this one... (Stop the service, copy over the existing files, restart the service), or if first time installation, extract into "C:\program files\glTF Thumbnail Provider" and run the install command.

I've managed to implement a windowless OpenTK session (using very basic OpenGL triangles at the moment), but at least it gives me a z-buffer.
It is considerably faster, and removes a lot of artifacts caused by the Windows GDI implementation.

Let me know how it goes!


________________________________

**Comment by vpenades** [comment]
@CentaurMare I've been trying to write a thumbnail service too, also by using SharpShell, but I fould a roadblock:  It seems the Shell API to use is SharpThumbnailHandler and implement IInitializeWithStream.... but this API only allows opening the main file.

With self contained GLB files, this is fine, but with glTF files that also have .BIN and textures, there's no way to reference them...  so how did you overcome this problem?

**Comment by Arlorean** [comment]
I think

**Comment by Arlorean** [comment]
Should we close off this issue now? I'd still like the option for a thumbnail but I can see that adding a custom extension to the Windows File Explorer is a viable alternative. Thanks for being open to discuss this.

**Comment by vpenades** [comment]
@Arlorean Unfortunately, using windows file explorer to add thumbnails is not a viable solution.

Windows provides an API for a service to generate thumbnails for a given file. Originally, the API allowed for both file paths and streams to be used to generate the thumbnails.

But in windows 10, the API that uses file paths has been locked for security reasons. The API still exists, but it fails if you try to use it.

This means that you're limited to use the API that gives you just a sanitized stream. The consequence is that you cannot resolve any file dependency, like loading the textures or the BIN blobs of a glTF file. At best you can generate thumbnails of stand alone GLB files.

@CentaurMare successfully achieved generating thumbnails thanks to extensive hacking skills, but I don't think it's the right approach for any safe thumbnail generation service.

So, unless Microsoft provides an alternative API that allows resolving file dependencies, right now it's extremely difficult to generate the thumbnails.

Furthermore: even if an extension to hold a thumbnail is implemented, the thumbnail content needs to be located in the root file...  so in case of glTF text files,  the thumbnail should be located within the json in embedded form, so the windows thumbnail service could retrieve it.

**Comment by emackey** [comment]
> even if an extension to hold a thumbnail is implemented, the thumbnail content needs to be located in the root file

For testing/development purposes, you don't need a whole extension.  Make a tiny JPG or PNG thumbnail-size image, encode it into a data-URI, and store that as a string in the JSON, under `asset.extras.thumbnail`.

This would bloat the thumbnail size a bit due to base64 encoding, but ideally the thumbnail should be quite small, and this meets the requirement to store it in the root JSON.
